src/Cardano/Address/Style/Shelley.hs

```diff
@@ -169,21 +222,74 @@ instance SoftDerivation Shelley where
             \either a programmer error, or, we may have reached the maximum \
             \number of addresses for a given wallet."
 
+instance HasNetworkDiscriminant Shelley where
+    type NetworkDiscriminant Shelley = NetworkTag
+    addressDiscrimination _ = RequiresNetworkTag
+    networkTag = id
+
+instance PaymentAddress Shelley where
+    paymentAddress discrimination k = unsafeMkAddress $
+        invariantSize expectedLength $ BL.toStrict $ runPut $ do
+            putWord8 firstByte
+            putByteString (blake2b224 k)
+      where
+          -- we use here the fact that payment address stands for what is named
+          -- as enterprise address, ie., address carrying no stake rights. For
+          -- rationale why we may need such addresses refer to delegation
+          -- specification - Section 3.2.3. What is important here is that the
+          -- address is composed of discrimination byte and 28 bytes hashed public key.
+          -- Moreover, it was decided that first 4 bits for enterprise address
+          -- will be `0110`. The next for 4 bits are reserved for network discriminator.
+          -- `0110 0000` is 96 in decimal.
+          firstByte =
+              96 + invariantNetworkTag (networkTag @Shelley discrimination)
```

Comment:
It'd be good to get a smart-constructor for `NetworkDiscriminant Shelley` that actually enforces this invariant. So that we can have: `mkNetworkDiscriminant :: Word8 -> Either XXX (NetworkDiscriminant Shelley)`. What do you think ?

Reply:
indeed, done in https://github.com/input-output-hk/cardano-addresses/pull/20/commits/78dd471eda7cf18172ea19b127165f8a2896c146